stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp

```diff
@@ -0,0 +1,457 @@
+using namespace tof::hardware;
+
+template <typename P>
+concept TMF8821Policy = requires(P p, uint16_t dev_addr, uint16_t reg,
+                                 uint16_t size, uint8_t* data) {
+    { p.i2c_read(dev_addr, reg, size) } -> std::same_as<RxTxReturn>;
+    { p.i2c_write(dev_addr, reg, data, size) } -> std::same_as<RxTxReturn>;
+};
+
+// I2C Device address
+constexpr uint16_t TOF_DEFAULT_ADDRESS = 0x41;
+constexpr uint16_t TOF_X_ADDRESS = 0x39;
+constexpr uint16_t TOF_Z_ADDRESS = 0x40;
+// Frame retry defaults
+constexpr uint8_t DEFAULT_RETRIES = 5;
+constexpr uint32_t DEFAULT_SLEEP_MS = 250;
+
+// Bootloader commands (bl_cmd_stat)
+
+// Remap RAM to Address 0 and Reset
+constexpr uint8_t BL_RAMREMAP_RESET = 0x11;
+// Initialize for RAM download from host to TMF8820/21/28
+constexpr uint8_t BL_DOWNLOAD_INIT = 0x14;
+// Build in self-test of RAM (pattern test)
+constexpr uint8_t BL_RAM_BIST = 0x2A;
+// Build in self-test of I²C RAM (pattern test)
+constexpr uint8_t BL_I2C_BIST = 0x2C;
+// Write RAM Region (Plain = not encoded into e.g. Intel Hex Records)
+constexpr uint8_t BL_W_RAM = 0x41;
+// Set the read/write RAM pointer to a given address
+constexpr uint8_t BL_ADDR_RAM = 0x43;
+
+// Bootloader cmd status
+enum BL_CMD_STATUS {
+    STAT_READY = 0,
+    STAT_ERR_SIZE,
+    STAT_ERR_CSUM,
+    STAT_ERR_RANGE,
+    STAT_ERR_MORE,
+};
+
+// Bootloader Payload constants
+constexpr uint8_t BL_DATA_LEN = 128;  // chunk len
+constexpr uint8_t BL_HEADER_LEN = 2;  // cmd + len
+constexpr uint8_t BL_FOOTER_LEN = 1;  // checksum
+constexpr uint8_t BL_BUFFER_LEN = BL_HEADER_LEN + BL_DATA_LEN + BL_FOOTER_LEN;
+uint8_t BUFFER[BL_BUFFER_LEN] = {0};
+
+// CMD_STAT Commands
+constexpr uint8_t CMD_WRITE_CONFIG_PAGE = 0x15;
+constexpr uint8_t CMD_LOAD_CONFIG_PAGE_COMMON = 0x16;
+constexpr uint8_t CMD_LOAD_CONFIG_PAGE_SPAD_1 = 0x17;
+constexpr uint8_t CMD_LOAD_CONFIG_PAGE_SPAD_2 = 0x18;
+constexpr uint8_t CMD_LOAD_CONFIG_PAGE_FACTORY_CALIB = 0x19;
+constexpr uint8_t CMD_FACTORY_CALIBRATION = 0x20;
+constexpr uint8_t CMD_I2C_SLAVE_ADDRESS = 0x21;
+constexpr uint8_t CMD_RESET = 0xFE;
+constexpr uint8_t CMD_STOP = 0xFF;
+
+// CMD_STAT Results
+constexpr uint8_t STAT_OK = 0x00;
+constexpr uint8_t STAT_ACCEPTED = 0x01;
+constexpr uint8_t STAT_ERR_CONFIG = 0x02;
+constexpr uint8_t STAT_ERR_APPLICATION = 0x03;
+constexpr uint8_t STAT_ERR_WAKEUP_TIMED = 0x04;
+constexpr uint8_t STAT_ERR_RESET_UNEXPECTED = 0x05;
+constexpr uint8_t STAT_ERR_UNKNOWN_CMD = 0x06;
+constexpr uint8_t STAT_ERR_NO_REF_SPAD = 0x07;
+constexpr uint8_t STAT_ERR_UNKNOWN_CID = 0x09;
+constexpr uint8_t STAT_WARNING_CONFIG_SPAD_1_NOT_ACCEPTED = 0x0A;
+constexpr uint8_t STAT_WARNING_CONFIG_SPAD_2_NOT_ACCEPTED = 0x0B;
+constexpr uint8_t STAT_WARNING_OSC_TRIM_NOT_ACCEPTED = 0x0C;
+constexpr uint8_t STAT_WARNING_I2C_ADDRESS_NOT_ACCEPTED = 0x0D;
+constexpr uint8_t STAT_ERR_UNKNOWN_MODE = 0x0E;
+
+class TMF8821 {
+  public:
+    TMF8821(TOFSensorPolicy* policy) : _policy(policy) {}
```

Comment:
why do we need the policy provided in both the ctor and the `initialize` function? 

also we should make any one-argument ctor `explicit`

Reply:
The policy in the ctor is initialized as a nullptr, then passed in the `initialize` function to set the pointer to the actual policy instance that we can reference within the class. Would it be better to pass in policy at compile time as a template parameter?

Reply:
removed from ctor since initializing as nullptr.